### ISSUE TITLE
how train vgg16 on my own data

### ISSUE THREAD

**Issue by BrettLL** [jcjohnson/fast-neural-style#95]
Hello,
was the provided model vgg16.t7 pretrained on the Imagenet?
If I want to train this model on my own data to match my own test data, how should  I do? Thank you~~

**Comment by htoyryla** [comment]
There is a long thread on this on the issues for the original neural-style https://github.com/jcjohnson/neural-style/issues/292

Before you start training your model, it might be a good idea to ask yourself what exactly you want to achieve. A good model for neural-style (including this version) needs to respond well to any essential features in both content and style images. The VGG16 and VGG19 models trained on Imagenet have been found to work well with quite varying materials. It would be hard to train a model that works better... even if one is using only special kinds of images.

I used to experiment training a VGG model using mainly geometric shapes, but came up with nothing useful. First of all, getting a deep convnet like a VGG trained well is difficult, and second, a network that responds to a limited set of features might not be a good idea anyway.

If you really think you have a useful dataset of your own, and it really is extensive enough (hundreds of thousands of images with labels), then you can try to train a useful VGG of your own. You should be prepared to experiment with settings and hyperparameters... it is not easy to get a deep convnet to start learning. You might need to train a less deep model first, and add layers on top of the pretrained layers to get a full depth net trained. 

If you just have a few thousands of images, it is a waste of time to try to train VGG model from scratch. It can be a useful learning exercise (for me it was) but the results will be poor by any standard. What you could do though, is to take a pretrained VGG and finetune it with your own images. Just remember that while finetuning is usually applied only to the FC layers to retrain the classifier only, for style transfer you need the conv layers (and you need to make sure that learning rate is non-zero for those layers). In theory, one could hope that the model then learns to adapt to the new features specific to the new material while at the same time keeping the performance regarding the already learned features. My experience, unfortunately, is that finetuning the conv layers of a pretrained VGG tend rather to mess up the performance than improve it. 

Still, training a VGG with a special dataset can be done. There exist pretrained models trained with photos of places (which I find works well with the original neural-style) and faces (which was a disappointment for me... apparently the model had learned to recognize the person but it really internalized how to detect generally useful facial features).

**Comment by filmo** [comment]
While I agree that retraining VGG seems like a lot of effort for little reward, it seems like using a particular set of image rather than the default COCO dataset might be worth exploring. For example, training a style image that is a portrait against a set of images that are only people might create a better model than one that's trained against COCO which has multiple different types of classes. Of course the final model would be more fragile with respect to any generic image, but might perform better against 'like-type' input images. 

Haven't tried that yet. Just speculation on my part having now trained a few models against the entire COCO data set.

**Comment by htoyryla** [comment]
I agree about using custom dataset for training the feed-forward model; I usually either use my own dataset, or sometimes I crosscheck using both my own and COCO.

In texture_nets, which is another feed-forward style transfer package, I usually use a dozen selected images, sometimes a single image. It is easy there the dataset is simply images in directories. According to the author of texture_nets, Dmitry Ulyanov, the best results are often obtained with a small dataset. 

Just in case for any reader who is not aware of the difference: the VGG model is almost like a standard used by almost all style transfer programs to evaluate content and style features. Fast-neural-style needs it in the process when you train your own feed-forward model.

**Comment by DirtyHarryLYL** [comment]
I noticed that the VGG is a "fixed"(paper) loss network during training. Does it mean that in backward() the weights of VGG won't update? 
"local grad_out_percep = percep_crit:backward(out, target)"?
"model:backward(x, grad_out)"
Best regards~
